### ISSUE TITLE
Bug: Node id collisions when inserting multiple documents

### ISSUE THREAD

**Issue by cdboer** [DLR-SC/gitlab2prov#34]
There seem to be merge issues when inserting documents that have node id's that are already given to nodes in the neo4j instance. This suggests, that the assumption of nodes being different when they are not in the same bundle but do have the same id is incorrect.

Approach to fix:
- include project/bundle id in each node id for nodes of a bundle

Create different id namespaces for different bundles.

**Comment by cdboer** [comment]
Fixed for now, data sharding will avoid this more elegantly. #35

**Comment by B-Stefan** [comment]
@cdboer  Do you have some example documents to reproduce this in the prov-db-connector project?

**Comment by cdboer** [comment]
> @cdboer Do you have some example documents to reproduce this in the prov-db-connector project?

Sorry for the late response, here are a few with a bit more explanation than the somewhat cryptic issue description provides.
Here is a link to a gist that can reproduce the issue: [Gist](https://gist.github.com/cdboer/3f0ca72a5153cad11e9ac2ce8695ac58)

I encountered this issue when trying to save multiple documents in one database.
Consider the following two independent documents:

**DocA**
```rdf
@prefix prov: <http://www.w3.org/ns/prov#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<docAbundleA> {
    <docAfoo> a prov:Agent,
            "user"^^xsd:string .
}
```

**DocB**
```rdf
@prefix prov: <http://www.w3.org/ns/prov#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<docBbundleB> {
    <docBfoo> a prov:Agent,
            "organization"^^xsd:string .
}
```
Both contain a Bundle with mutually different names.
The bundles themselves both contain an agent with the identifier `foo`.

In DocA agent `foo` represents a user and therefore has prov:type "user".
In DocB agent `foo` represents an organization and bares the appropriate prov:type "organization".

When inserting both documents into the same empty database the following exception occurs.

```
provdbconnector.exceptions.database.MergeException: The attributes [<QualifiedName: prov:type>, 'meta:prov_type', 'meta:identifier', 'meta:namespaces', 'meta:type_map', 'meta:aae65ff5-7662-4ac2-a374-485e218a3a93'] could not merged into the existing node, All attributes: {'prov:type': 'organization', 'meta:prov_type': 'prov:Agent', 'meta:identifier': 'foo', 'meta:namespaces': '{"prov": "http://www.w3.org/ns/prov#", "": "docB"}', 'meta:type_map': '{}', 'meta:aae65ff5-7662-4ac2-a374-485e218a3a93': True}
```

I expected that nodes of different documents that are additionally denoted in different bundles to be different no matter their id. So this caught me offguard.

When trying the same with documents that contain bundles with agents that share the same id aswell as label information the nodes get merged.

**DocA**
```rdf
@prefix prov: <http://www.w3.org/ns/prov#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<docAbundleA> {
    <docAfoo> a prov:Agent,
            "organization"^^xsd:string .
}
```

**DocB**
```rdf
@prefix prov: <http://www.w3.org/ns/prov#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<docBbundleB> {
    <docBfoo> a prov:Agent,
            "organization"^^xsd:string .
}
```

Let's check how many different agents exist:

```bash
echo "MATCH (a:Agent) RETURN COUNT(a) AS agentCount;" | cypher-shell --format verbose
+------------+
| agentCount |
+------------+
| 1          |
+------------+

1 row available after 9 ms, consumed after another 0 ms
```

I am not shure whether this is intended behaviour but maybe you can help me out.

**Comment by B-Stefan** [comment]
@cdboer  Thanks for the detailed information! 
I am pretty confident that you found a bug. Unfortunately, I'm quite busy this week so I will look into this during the next weekend.

I noticed that you used different default namespaces for the documents in your gist. So the full qualified name/identifier of `docAfoo` would be `docA: foo`  and for B `dockB:foo`. That should do the trick to separate documents from each other.

**Comment by B-Stefan** [comment]
@cdboer  The good news: I found the issue; you discovered an architectural flaw in the lib 👍 
 
Do you use custom Cypher commands to analyze the generated provenance data in the neo4j database?  The solution for this issue will cause a breaking change for existing Cypher queries.

**Comment by cdboer** [comment]
Yes, i do. Though only a handful, would not be too much work to reconstruct them after the breaking changes.

**Comment by B-Stefan** [comment]
Fixed see version v0.4.0 - https://github.com/DLR-SC/prov-db-connector/releases/tag/v0.4.0

**Comment by cdboer** [comment]
Great news, thanks a lot!
